core/store/src/trie/mem/lookup.rs

```diff
@@ -1,37 +1,39 @@
+use super::metrics::MEM_TRIE_NUM_LOOKUPS;
 use super::node::{MemTrieNodePtr, MemTrieNodeView};
 use crate::NibbleSlice;
-use near_primitives::hash::CryptoHash;
+use near_primitives::hash::{hash, CryptoHash};
 use near_primitives::state::FlatStateValue;
-use near_vm_runner::logic::TrieNodesCount;
-use std::collections::HashSet;
 
 /// Performs a lookup in an in-memory trie, while taking care of cache
 /// accounting for gas calculation purposes.
+///
+/// The provided callback is invoked for:
+///  - each node accessed during the lookup, with the parameters
+///    (false, node hash, borsh serialized node), as well as
+///  - once for the value, if the value is inlined, with the parameters
+///    (true, value hash, the value).
+///
+/// The callback is used for:
+///  - Gas accounting of touched nodes and values
+///  - Populating the trie accounting cache (for legacy compatibility)
+///  - Recording accessed nodes and values, for state witness production.
 pub fn memtrie_lookup(
     root: MemTrieNodePtr<'_>,
     key: &[u8],
-    mut accessed_cache: Option<&mut HashSet<CryptoHash>>,
-    nodes_count: &mut TrieNodesCount,
+    mut node_access_callback: impl FnMut(bool, CryptoHash, Vec<u8>),
```

Comment:
I think it complicates logic a lot. How about we make this function optionally return list of triplets and let caller handle it?

Reply:
Oh yeah, that sounds like a good idea. Actually it probably isn't even necessary to return a triplet coz the value access can be derived from the return value...